external/fv3fit/fv3fit/reservoir/domain.py

```diff
@@ -0,0 +1,76 @@
+import numpy as np
+
+
+def _slice(arr: np.ndarray, inds: slice, axis: int = 0):
+    # https://stackoverflow.com/a/37729566
+    # For slicing ndarray along a dynamically specified axis
+    # same as np.take() but does not make a copy of the data
+    sl = [slice(None)] * arr.ndim
+    sl[axis] = inds
+    return arr[tuple(sl)]
+
+
+class Subdomain:
+    def __init__(self, data: np.ndarray, overlap: int, subdomain_axis: int = 0):
+        self.overlapping = data
+        self.overlap = overlap
+        self.nonoverlapping = _slice(
+            arr=data, inds=slice(overlap, -overlap), axis=subdomain_axis
+        )
+
+
+class PeriodicDomain:
+    def __init__(
+        self, data: np.ndarray, output_size: int, overlap: int, subdomain_axis: int = 0
+    ):
+        self.data = data
+        self.output_size = output_size
+        if data.shape[subdomain_axis] % output_size != 0:
+            raise ValueError(f"Data size must be evenly divisible by output_size")
+        self.overlap = overlap
+        self.subdomain_axis = subdomain_axis
+        self.n_subdomains = int(data.shape[subdomain_axis] / output_size)
+        self.index = 0
+
+    def __len__(self) -> int:
+        return self.n_subdomains
+
+    def __getitem__(self, index: int):
+        padded = np.hstack(
+            [
+                _slice(
+                    arr=self.data,
+                    inds=slice(-self.overlap, None),
+                    axis=self.subdomain_axis,
+                ),
+                self.data,
+                _slice(
+                    arr=self.data,
+                    inds=slice(None, self.overlap),
+                    axis=self.subdomain_axis,
+                ),
+            ]
+        )
```

Comment:
Thanks, I didn't know about those numpy functions. `mode="wrap"` in `np.pad` is neat functionality, but there's no option to only pad a single axis. `np.roll` is also a neat way to wrap the values, though it would still require some kind of slicing and appending of the rolled matrices like what is done here.

An alternative to simplifying some of this slicing code would be to require the input  arrays to always be 2 dimensional with the first dimension being time (i.e. in the prediction case reshape to (1, -1)). This might be less future-proof though if we want to reuse 1D subdomains in the future for vertical levels. Let me know if you'd prefer that.

Reply:
It's admittedly a bit clunky, but with `np.pad` you can specify axes by providing a tuple of tuples for the `pad_width`:
```
>>> import numpy as np
>>> arr = np.array([[1, 2, 3], [4, 5, 6]])
>>> np.pad(arr, ((0, 0), (1, 1)), mode="wrap")
array([[3, 1, 2, 3, 1],
       [6, 4, 5, 6, 4]])
```
With `np.roll` the way I was thinking could avoid concatenating (obviously slicing is still required).  If we roll the array along the axis such that the subdomain origin is at the index `overlap`, then selecting the data becomes a matter of selecting the first `2 * overlap + subdomain_size` values along the axis:
```
>>> arr = np.arange(16)
>>> subdomain_index = 1; overlap = 2; subdomain_size = 4
>>> shift = -(subdomain_index * subdomain_size + overlap)
>>> np.roll(arr, shift)[:2 * overlap + subdomain_size]
array([ 6,  7,  8,  9, 10, 11, 12, 13])
```
I think `hstack` inherently makes an assumption about the axis that you would like to concatenate along, so one of those approaches might be more robust.

Reply:
Thanks, I didn't know about the tuple option for `pad_width`- that's really useful.